webrender/src/batch.rs

```diff
@@ -442,43 +633,53 @@ impl AlphaBatchContainer {
         self.alpha_batches.is_empty()
     }
 
-    fn merge(&mut self, builder: AlphaBatchBuilder, task_rect: &DeviceIntRect) {
-        self.task_rect = self.task_rect.union(task_rect);
-
-        for other_batch in builder.opaque_batch_list.batches {
-            let batch_index = self.opaque_batches.iter().position(|batch| {
-                batch.key.is_compatible_with(&other_batch.key)
-            });
-
-            match batch_index {
-                Some(batch_index) => {
-                    self.opaque_batches[batch_index].merge(other_batch);
-                }
-                None => {
-                    self.opaque_batches.push(other_batch);
-                }
+    // Merge the `from_list` into the `to_list` by preserving the order of batches on both sides
+    // while placing compatible batch keys together.
+    fn consume_list(to_list: &mut Vec<PrimitiveBatch>, from_list: Vec<PrimitiveBatch>, stack_index_offset: StackIndex) {
+        to_list.extend(
+            from_list.iter().map(|_| PrimitiveBatch::new(BatchKey::DUMMY, !0, 0))
+        );
+        to_list.rotate_right(from_list.len());
+        let (mut src_batch_index, mut dst_batch_index) = (from_list.len(), 0);
+
+        for mut other_batch in from_list {
+            while to_list
+                .get(src_batch_index)
+                .map_or(false, |b| !b.key.is_compatible_with(&other_batch.key))
+            {
+                to_list[dst_batch_index] = to_list[src_batch_index].clone();
+                src_batch_index += 1;
+                dst_batch_index += 1;
             }
+            other_batch.stack_index += stack_index_offset;
+            to_list[dst_batch_index] = other_batch;
+            dst_batch_index += 1;
         }
 
-        let mut min_batch_index = 0;
+        assert_eq!(src_batch_index, dst_batch_index);
+    }
 
-        for other_batch in builder.alpha_batch_list.batches {
-            let batch_index = self.alpha_batches.iter().skip(min_batch_index).position(|batch| {
-                batch.key.is_compatible_with(&other_batch.key)
-            });
+    fn merge(&mut self, builder: AlphaBatchBuilder, task_rect: &DeviceIntRect) {
+        self.task_rect = self.task_rect.union(task_rect);
 
-            match batch_index {
-                Some(batch_index) => {
-                    let index = batch_index + min_batch_index;
-                    self.alpha_batches[index].merge(other_batch);
-                    min_batch_index = index;
-                }
-                None => {
-                    self.alpha_batches.push(other_batch);
-                    min_batch_index = self.alpha_batches.len();
-                }
-            }
-        }
+        // Note: we are able to reorder the opaque batches as needed, theoretically.
```

Comment:
I don't think this will have any effect on fill rate. By definition, batch merging only occurs for non-overlapping regions of a target (separate render tasks). Does that have an effect on what we should do here by default?